### PR TITLE
Refs [TASK] [36130] Handle Detail Champion

### PR DESCRIPTION
**Pull request by thanhvinh2k**
## Related Tickets
- [#Task 36130: Click hero item to detail champion](https://edu-redmine.sun-asterisk.vn/issues/36130)
## WHAT
- Click item to detail champion
## Evidence (Screenshot or Video)
heroesScreen
detailScreen

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sun/hero_01/utils/OnItemHeroClickListener.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.sun.hero_01.utils
+
+interface OnItemHeroClickListener<T> {
+    fun onItemClickListener(itemHero: T?)
+}
```
Comment (khoaqh-2470) [inline app/src/main/java/com/sun/hero_01/utils/OnItemHeroClickListener.kt]:
```suggestion
interface OnItemHeroClickListener<T> {
    fun onItemClickListener(itemHero: T?)
}

```

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt [hunk 2]
```diff
@@ -9,16 +9,17 @@ import com.sun.hero_01.data.model.Hero
...
 import com.sun.hero_01.utils.HeroDifficulty
 import com.sun.hero_01.utils.LoadImageBitmap
+import com.sun.hero_01.utils.OnItemHeroClickListener
 import kotlinx.android.synthetic.main.item_layout_hero.view.*
 
 class ChampionAdapter : RecyclerView.Adapter<ChampionAdapter.ViewHolder>() {
 
     private val heroes = mutableListOf<Hero>()
-
+    private var onItemClickListener: OnItemHeroClickListener<Hero>? = null
```
Comment (khoaqh-2470) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt]: suggest higherorder function

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 3]
```diff
@@ -53,4 +60,8 @@ class ChampionFragment : BaseFragment(), ChampionContract.View {
     companion object {
         fun newInstance() = ChampionFragment()
     }
+
+    override fun onItemClickListener(itemHero: Hero?) {
+        this.replaceFragment(DetailFragment.newInstance(itemHero?.name!!), R.id.frameContainer)
+    }
```
Comment (khoaqh-2470) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: Sắp xếp lại các function theo thứ tự override -> private

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt (hunk 2 above)
Comment (datdt-1935) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt]: để lên constructor cho gọn nhé

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt [hunk 4]
```diff
@@ -45,13 +50,18 @@ class ChampionAdapter : RecyclerView.Adapter<ChampionAdapter.ViewHolder>() {
         }
     }
 
-    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+    inner class ViewHolder(itemView: View, private val itemListener: OnItemHeroClickListener<Hero>?)
```
Comment (datdt-1935) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt]: break line

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt [hunk 5]
```diff
@@ -45,13 +50,18 @@ class ChampionAdapter : RecyclerView.Adapter<ChampionAdapter.ViewHolder>() {
...
+        : RecyclerView.ViewHolder(itemView), View.OnClickListener {
+
+        private var listener: OnItemHeroClickListener<Hero>? = null
 
         fun bindViewData(hero: Hero) {
             itemView.apply {
                 textViewHeroName.text = hero.name
                 textViewHeroNickname.text = hero.title
                 textViewHeroDifficulty.text = getDifficulty(hero.difficulty)
+                itemView.setOnClickListener(this@ViewHolder)
```
Comment (datdt-1935) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionAdapter.kt]: để vào init để tránh lặp lại mỗi lần bindData

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 6]
```diff
@@ -53,4 +60,8 @@ class ChampionFragment : BaseFragment(), ChampionContract.View {
     companion object {
         fun newInstance() = ChampionFragment()
     }
+
+    override fun onItemClickListener(itemHero: Hero?) {
+        this.replaceFragment(DetailFragment.newInstance(itemHero?.name!!), R.id.frameContainer)
```
Comment (datdt-1935) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: không assert non-null

File: app/src/main/java/com/sun/hero_01/ui/detail/DetailFragment.kt [hunk 7]
```diff
@@ -42,7 +42,7 @@ class DetailFragment : BaseFragment(), DetailContract.View {
     companion object {
         private const val ARGUMENT_HERO_ID = "ARGUMENT_HERO_ID"
 
-        fun newInstance(heroName: String) =  DetailFragment().apply{
+        fun newInstance(heroName: String?) =  DetailFragment().apply{
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/detail/DetailFragment.kt]:
```suggestion
        fun newInstance(heroName: String?) =  DetailFragment().apply {
```

File: app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt [hunk 8]
```diff
@@ -1,18 +1,24 @@
 package com.sun.hero_01.ui.champion
 
 import android.os.Bundle
+import android.util.Log
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/ui/champion/ChampionFragment.kt]: optimize imports

File: app/src/main/java/com/sun/hero_01/utils/OnItemHeroClickListener.kt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package com.sun.hero_01.utils
+
+interface OnItemHeroClickListener<T> {
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/sun/hero_01/utils/OnItemHeroClickListener.kt]:
```suggestion
interface OnItemRecyclerViewListener<T> {
```

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**